File: src/main/java/com/mythicalcreaturesoftware/splash/filereader/FileReaderFactory.java
package com.mythicalcreaturesoftware.splash.filereader;

import com.mythicalcreaturesoftware.splash.exception.UnsupportedFileTypeException;
import com.mythicalcreaturesoftware.splash.filereader.impl.CbrFileReader;
import com.mythicalcreaturesoftware.splash.filereader.impl.CbzFileReader;

public class FileReaderFactory {

    public static FileReader buildFileReader(FileReaderType type, String path) throws UnsupportedFileTypeException {

        FileReader fileReader;

        switch (type) {
            case CBR:
                fileReader = new CbrFileReader(path);
                break;

            case CBZ:
                fileReader = new CbzFileReader(path);
                break;

            default:
                throw new UnsupportedFileTypeException("Unsupported File Type: " + type);
        }

        return fileReader;
    }
}

File: reader/src/commonMain/kotlin/reader/filereader/FileReaderFactory.kt
package reader.filereader

import reader.exception.UnsupportedFileTypeException
import reader.filereader.impl.CbrFileReader
import reader.filereader.impl.CbzFileReader

/**
 * Receives a [FileReaderType] with name [type] and a [String] with name [path], containing the path of a file.
 * Returns the respective implementation of [FileReader] for the given type
 */
fun buildFileReader(type: FileReaderType, path: String ): FileReader {
    val reader = when (type) {
        FileReaderType.CBR -> CbrFileReader(path)
        FileReaderType.CBZ -> CbzFileReader(path)
        else -> {
            throw UnsupportedFileTypeException("Unsupported File Type: $type")
        }
    }

    reader.initialize()

    return reader
}
